yacl/crypto/primitives/tpre/tpre.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 Chengfang Financial Technology Co., Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/tpre/tpre.h"
+
+#include <vector>
+
+#include "yacl/crypto/base/aead/sm4_mac.h"
+#include "yacl/crypto/primitives/tpre/capsule.h"
+
+namespace yacl::crypto {
+
+std::pair<Capsule::CapsuleStruct, std::vector<uint8_t>> TPRE::Encrypt(
+    const std::unique_ptr<EcGroup>& ecc_group, const Keys::PublicKey& pk_A,
+    const std::string& iv, const std::string& plaintext) const {
+  Capsule capsule;
+
+  std::pair<Capsule::CapsuleStruct, std::vector<uint8_t>> capsule_pair =
+      capsule.EnCapsulate(ecc_group, pk_A);
+
+  std::vector<uint8_t> ciphertext =
+      yacl::crypto::Sm4MteEncrypt(capsule_pair.second, iv, plaintext);
+
+  return {capsule_pair.first, ciphertext};
+}
+
+std::string TPRE::Decrypt(const std::unique_ptr<EcGroup>& ecc_group,
+                          const Capsule::CapsuleStruct& capsule_struct,
+                          const std::string& iv,
+                          const std::vector<uint8_t>& enc_data,
+                          const Keys::PrivateKey& sk_A) const {
+  Capsule capsule;
+  std::vector<uint8_t> dek =
+      capsule.DeCapsulate(ecc_group, sk_A, capsule_struct);
+
+  std::string dek_str(dek.begin(), dek.end());
+  std::vector<uint8_t> plaintext =
+      yacl::crypto::Sm4MteDecrypt(dek_str, iv, enc_data);
```

Comment:
Could directly use `dek`, not need generate a temp var `dek_str `. Same reason as for `SM3`.